src/Formats/IonWriter.cpp

```diff
@@ -0,0 +1,355 @@
+    // flash is called only if we are at the zero nesting level, that is, not in a list or structure,
+    // which means one full ion message will be saved to the internal buffer,
+    // and then it will be transferred to WriteBuffer
+    ion_writer_callback = [this](_ion_user_stream * stream) -> iERR
+    {
+        // if limit is init it is call from flush
+        if (stream->limit == ion_buffer.data() + ion_buffer.size())
+        {
+            // Save current total written bytes
+            if ((stream->curr >= ion_buffer.data()) && (stream->curr <= ion_buffer.data() + ion_buffer.size()))
+                total_written_bytes = stream->curr - ion_buffer.data();
+
+            // Checking whether the buffer has overflowed; if so, expand and update the stream pointers
+            if (stream->curr == stream->limit)
+            {
+                // todo: Should we put a border on top of this?
+                ion_buffer.resize(ion_buffer.size() << 1);
+                stream->limit = ion_buffer.data() + ion_buffer.size();
+                stream->curr = ion_buffer.data() + total_written_bytes;
+            }
+        }
+        else
+        {
+            // else it is first call
+            total_written_bytes = 0;
+            stream->limit = ion_buffer.data() + ion_buffer.size();
+            stream->curr = ion_buffer.data();
+        }
+        return IERR_OK;
+    };
+
+    // Init stream and writer
+    ION_STREAM * stream = nullptr;
+    hWRITER writer = nullptr;
+    ION_CHECK(ion_stream_open_handler_out(&ionWriteCallback, &ion_writer_callback, &stream), "Cannot create Ion format writer")
+    auto err = ion_writer_open(&writer, stream, &ion_options);
+    // IERR_OK == 0
+    if (err)
+    {
+        ion_stream_close(stream);
+        throw Exception(ErrorCodes::ION_FORMAT_WRITER_ERROR, "Cannot create ION format writer, error: {}", ion_error_to_str(err));
+    }
+
+    ion_stream = stream;
+    ion_writer = writer;
+}
+
+IonWriter::~IonWriter()
+{
+    if (ion_writer != nullptr)
+    {
+        ion_writer_close(ion_writer);
+        ion_stream_close(ion_stream);
+        ion_writer = nullptr;
+        ion_stream = nullptr;
+    }
+    // TODO: Is it necessary to somehow handle errors from ion-c here?
+    // ion_writer: if any value is in-progress, closing any writer raises an error, but still frees the writer and any associated memory.
+    // (only one possible error)
+    // ion_stream: Possible errors from write during the final flush, but if we used callback handler only error from it is possible
+    //    if (ion_writer != nullptr)
+    //    {
+    //        ION_CHECK(ion_writer_close(ion_writer), "Cannot close Ion writer")
+    //        ion_writer = nullptr;
+    //    }
+    //    if (ion_stream != nullptr)
+    //    {
+    //        ION_CHECK(ion_stream_close(ion_stream), "Cannot close Ion stream")
+    //        ion_stream = nullptr;
+    //    }
+}
```

Comment:
It's better not to do any work in the destructor. We can move it to a separate method like `close` and call it in `final`. And we can handle errors from ion-c in that case. And also we will call `close` in try/catch in destructor only if it wasn't called explicitely.